### PR TITLE
feat: Make Yaml nodes serializable

### LINKED ISSUE #641: Make YamlNode serializable
### Describe the problem you'd like to solve

I would like to deserialize structures, that are partly arbitrary. For example when I want to deserialize the [mustache specification], most of the structure can be properly deserialized with somethling like
```kotlin
@Serializable
data class Specification(
  val overview: String,
  val tests: List<Test>,
)

@Serializable
data class Test(
  val name: String,
  val desc: String,
  val template: String,
  val expected: String,
  val partials: Map<String, String> = emptyMap(),
  val data: YamlNode? = null, // not serializable
)
```
but the `data` can't be, because it can be arbitrary nested or even just a primitive value.

### Describe the solution you'd like

I would like to have `YamlNode` as `Serializable` (similar as done by the kotlinx team for the [`JsonElement`](https://github.com/Kotlin/kotlinx.serialization/blob/d4d066d72a9f92f06c640be5a36a22f75d0d7659/formats/json/commonMain/src/kotlinx/serialization/json/JsonElement.kt#L26)).

### Describe alternatives you've considered

I could write it on my own in my project, but I think, this would be really valuable to the library. Another alternative is to read the whole structure as `YamlNode` and then parse everything on with the node, but that reduces the usability.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by EdwarDDay**
- fixes #641

### CODE REVIEW

File: src/commonMain/kotlin/com/charleskorn/kaml/YamlNodeSerializer.kt [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+import kotlinx.serialization.descriptors.PrimitiveSerialDescriptor
+import kotlinx.serialization.descriptors.SerialDescriptor
+import kotlinx.serialization.descriptors.SerialKind
+import kotlinx.serialization.descriptors.buildSerialDescriptor
+import kotlinx.serialization.descriptors.nullable
+import kotlinx.serialization.encoding.Decoder
+import kotlinx.serialization.encoding.Encoder
+import kotlinx.serialization.encoding.encodeStructure
+
+internal object YamlNodeSerializer : YamlContentPolymorphicSerializer<YamlNode>(YamlNode::class) {
```
Comment (charleskorn) [inline src/commonMain/kotlin/com/charleskorn/kaml/YamlNodeSerializer.kt:42]: Can't this just be an ordinary `KSerializer`?

File: src/commonMain/kotlin/com/charleskorn/kaml/YamlNodeSerializer.kt [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+        } catch (_: SerializationException) {
+        }
+        try {
+            return encoder.encodeLong(value.toLong())
+        } catch (_: SerializationException) {
+        }
+        try {
+            return encoder.encodeDouble(value.toDouble())
+        } catch (_: SerializationException) {
+        }
```
Comment (charleskorn) [inline src/commonMain/kotlin/com/charleskorn/kaml/YamlNodeSerializer.kt:83]: Rather than this chain of `try` / `catch`, couldn't we inspect the value and determine the best option to use?

File: src/commonMain/kotlin/com/charleskorn/kaml/YamlInput.kt [hunk 3]
```diff
@@ -62,6 +62,13 @@ public sealed class YamlInput(
             is YamlList -> when (descriptor.kind) {
                 is StructureKind.LIST -> YamlListInput(node, yaml, context, configuration)
                 is SerialKind.CONTEXTUAL -> createContextual(node, yaml, context, configuration, descriptor)
+                is PolymorphicKind -> {
+                    if (descriptor.isContentBasedPolymorphic) {
+                        createContextual(node, yaml, context, configuration, descriptor)
+                    } else {
+                        throw MissingTypeTagException(node.path)
+                    }
+                }
```
Comment (charleskorn) [inline src/commonMain/kotlin/com/charleskorn/kaml/YamlInput.kt:71]: Why is this change required?

File: src/commonTest/kotlin/com/charleskorn/kaml/YamlWritingTest.kt [hunk 4]
```diff
@@ -729,6 +793,46 @@ class YamlWritingTest : FlatFunSpec({
...
+                      - 2
+                      - 3
+                """.trimIndent(),
+            ) as YamlMap
+            val expectedOutput = """
+                    text: "test"
+                    node:
+                      "'foo'": "'bar'"
+                      "'baz'": 1
+                      "'test'":
```
Comment (charleskorn) [inline src/commonTest/kotlin/com/charleskorn/kaml/YamlWritingTest.kt:812]: Why are the values quoted twice here?

**Review by charleskorn — COMMENTED** [review]
Thanks for the PR @EdwarDDay!

File: src/commonMain/kotlin/com/charleskorn/kaml/YamlNodeSerializer.kt (hunk 1 above)
Comment (EdwarDDay) [reply src/commonMain/kotlin/com/charleskorn/kaml/YamlNodeSerializer.kt]: Changed. I mainly did it, because the `Marker` annotation is needed for the polymorphic handling.

File: src/commonMain/kotlin/com/charleskorn/kaml/YamlNodeSerializer.kt (hunk 2 above)
Comment (EdwarDDay) [reply src/commonMain/kotlin/com/charleskorn/kaml/YamlNodeSerializer.kt:81]: I didn't want to copy the parsing code, to avoid duplicated code. Do you have an idea how to implement this without duplicating most of the `YamlScalar` code?
I mean `Boolean` has extra parsing code, integer values extra checks for octal, decimal and hexadecimal numbers, floating point numbers have extra checks for infinity and NaN. This was too much code for me to copy. Else the library gets inconsistent with future changes.

File: src/commonMain/kotlin/com/charleskorn/kaml/YamlInput.kt (hunk 3 above)
Comment (EdwarDDay) [reply src/commonMain/kotlin/com/charleskorn/kaml/YamlInput.kt:71]: To be able to decode a `YamlList` still as `YamlNode`. The `YamlNodeSerializer.descriptor` is polymorphic (because we don't know the type of it). So a list needs to be readable as polymorphic.

File: src/commonTest/kotlin/com/charleskorn/kaml/YamlWritingTest.kt (hunk 4 above)
Comment (EdwarDDay) [reply src/commonTest/kotlin/com/charleskorn/kaml/YamlWritingTest.kt]: Oh, that was an error. Fixed it.

File: src/commonMain/kotlin/com/charleskorn/kaml/YamlNodeSerializer.kt [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+        value.content.singleOrNull()?.also { return encoder.encodeChar(it) }
+        encoder.encodeString(value.content)
```
Comment (charleskorn) [inline src/commonMain/kotlin/com/charleskorn/kaml/YamlNodeSerializer.kt:83]: Is there a test for this special handling of characters?

File: src/commonMain/kotlin/com/charleskorn/kaml/YamlNodeSerializer.kt (hunk 2 above)
Comment (charleskorn) [reply src/commonMain/kotlin/com/charleskorn/kaml/YamlNodeSerializer.kt:81]: What if we just always encode the value as a string?

Comment (EdwarDDay) [reply src/commonMain/kotlin/com/charleskorn/kaml/YamlNodeSerializer.kt:81]: Then we would encode a double `1.2` as `"1.2"`, right? So the output might be unexpected to the user.

I could create `toBooleanOrNull`, `toByteOrNull`, …  functions in the `YamlScalar` class and use it in `toBoolean`, `toByte`, … functions and in the `YamlScalarSerializer`. This way the code doesn't get duplicated and we don't have to try catch everything.

File: src/commonMain/kotlin/com/charleskorn/kaml/YamlNodeSerializer.kt (hunk 5 above)
Comment (EdwarDDay) [reply src/commonMain/kotlin/com/charleskorn/kaml/YamlNodeSerializer.kt:83]: No, I just wanted to use the short version of `encodeChar`, to not have the overhead from `encodeString`. There are also no extra tests for `encodeBoolean`, `encodeLong`, `encodeDouble`. I thought, these are properly tested in the tests for `YamlOutput`.
But I can of course write extra tests, if you think, that's needed.

Comment (charleskorn) [reply src/commonMain/kotlin/com/charleskorn/kaml/YamlNodeSerializer.kt:83]: I think it'd be valuable to have a test that shows the expected behaviour when a value looks like it is a boolean or a number etc.

File: src/commonMain/kotlin/com/charleskorn/kaml/YamlNodeSerializer.kt (hunk 2 above)
Comment (charleskorn) [reply src/commonMain/kotlin/com/charleskorn/kaml/YamlNodeSerializer.kt:81]: > I could create `toBooleanOrNull`, `toByteOrNull`, … functions in the `YamlScalar` class and use it in `toBoolean`, `toByte`, … functions and in the `YamlScalarSerializer`. This way the code doesn't get duplicated and we don't have to try catch everything.

This sounds good to me.

Comment (EdwarDDay) [reply src/commonMain/kotlin/com/charleskorn/kaml/YamlNodeSerializer.kt]: done

File: src/commonMain/kotlin/com/charleskorn/kaml/YamlNodeSerializer.kt (hunk 5 above)
Comment (EdwarDDay) [reply src/commonMain/kotlin/com/charleskorn/kaml/YamlNodeSerializer.kt]: I added some tests for encoding the primitives via the new serializer.

**Review by charleskorn — APPROVED** [review]
Thanks @EdwarDDay!
